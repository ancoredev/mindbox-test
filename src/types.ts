export interface ITodo {
  id: number,
  task: string,
  isCompleted: boolean
}

export type THandlerById = (id: number) => void;

export type TMuiClickhandler = (
  event: React.SyntheticEvent, 
  newValue: number
) => void;

export type TFilterTasks = (tasks: ITodo[], method: number) => ITodo[];

export type TTaskActions = {
  tasks: ITodo[],
  addTask: (taskValue: string) => void,
  removeTask: THandlerById,
  toggleTask: THandlerById,
  clearCompletedTasks: () => void
}