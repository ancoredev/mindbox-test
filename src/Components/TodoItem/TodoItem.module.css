.list-item {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  background-color: #ffffff;
  box-sizing: border-box;
  border-radius: 5px;
  padding: 5px 10px 5px 10px;
  width: 600px;
  margin-bottom: 5px;
  font-size: 18px;
}

.list-item p {
  flex-basis: 480px;
}

.list-item button {
  background-color: transparent;
  border: none;
  cursor: pointer;
  width: 40px;
  height: 40px;
  padding: 0;
}

