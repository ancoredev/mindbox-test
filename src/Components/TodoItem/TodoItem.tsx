import classes from './TodoItem.module.css'

import { FC } from 'react'
import { ITodo, THandlerById } from '../../types'

import { Checkbox } from '@mui/material'
import ClearIcon from '@mui/icons-material/Clear'


interface TodoItemProps extends ITodo {
  removeTask: THandlerById,
  toggleTask: THandlerById
}

const TodoItem: FC<TodoItemProps> = ({ id, task, isCompleted, removeTask, toggleTask }) => {
  return (
    <li className={classes['list-item']}>
      <Checkbox 
        checked={isCompleted}
        onChange={() => toggleTask(id)}
      />
      <p>{task}</p>
      <button
        onClick={() => removeTask(id)}
      ><ClearIcon/></button>
    </li>
  )
}

export default TodoItem