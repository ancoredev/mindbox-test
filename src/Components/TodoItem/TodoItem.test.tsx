import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ITodo } from '../../types';

import TodoItem from './TodoItem';


const taskExample: ITodo = { id: 2, task: 'Прекрасный код', isCompleted: true }

const toggleTask = jest.fn();
const removeTask = jest.fn();

const setup = ({
  id, 
  task, 
  isCompleted,
  }: ITodo): {
    liEl: HTMLLIElement,
    component: RenderResult
} => {
  const component = render(
    <TodoItem
      id={id}
      task={task}
      isCompleted={isCompleted}
      removeTask={removeTask}
      toggleTask={toggleTask}
    />); 
  const liEl = screen.getByRole<HTMLLIElement>('listitem');
  return {
    liEl,
    component
  }
}


describe('TodoItem component', () => {
  it('should render', () => {
    const { liEl} = setup(taskExample);

    expect(liEl).toBeInTheDocument();
  });

  it('should match the snapshot', () => {
    const { component } = setup(taskExample);
    expect(component).toMatchSnapshot();
  });

  it('should display correct task passed by prop', () => {
    setup({
      id: 1,
      task: 'test',
      isCompleted: false
    });

    expect(screen.getByRole('checkbox')).not.toBeChecked();
    expect(screen.getByText('test', { selector: 'p' })).toBeInTheDocument();
  });

  it('should correct call onChange and onClick events', async () => {
    const user = userEvent.setup();
    setup({
      id: 1,
      task: 'test',
      isCompleted: false
    });

    expect(screen.getByRole('checkbox')).not.toBeChecked();
    
    await user.click(screen.getByRole('checkbox'));
    expect(toggleTask).toHaveBeenCalled();

    await user.click(screen.getByRole('button'));
    expect(removeTask).toHaveBeenCalled();
  });
});