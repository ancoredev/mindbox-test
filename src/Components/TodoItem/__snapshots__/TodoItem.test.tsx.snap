// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem component should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="list-item"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <p>
          Прекрасный код
        </p>
        <button>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ClearIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="list-item"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          checked=""
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <p>
        Прекрасный код
      </p>
      <button>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ClearIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
